File: python/Remove-Linked-List-Elements/Remove-Linked-List-Elements.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        p = ListNode(-1)
        p.next = head
        prev = p
        cur = head
        while cur:
            if cur.val == val:
                prev.next = cur.next
            else:
                prev = cur
            cur = cur.next
        return p.next

File: cpp/Remove-Linked-List-Elements/Remove-Linked-List-Elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* p = new ListNode(-1);
        p->next = head;
        ListNode* prev = p;
        ListNode* cur = head;
        ListNode* del = NULL;
        while(cur)
        {
            if (cur->val==val)
            {
                prev->next = cur->next;
                del = cur;
            }
            else
            {
                prev = cur;
            }
            cur = cur->next;
            if (del)
            {
                delete del;
                del = NULL;
            }
        }
        ListNode* res = p->next;
        delete p;
        return res;
    }
};
